### PR TITLE
My first refactor

### PR DESCRIPTION
**Pull request by rudolfbyker**
- Less code duplication
- Some style improvements
- Progress towards https://github.com/lbdremy/solr-node-client/issues/251
- All tests are still passing, as far as I can tell

Next step after merge: Refactor to use promises everywhere.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,19 +61,26 @@ client.add({ id : 12, title_t : 'Hello' },function(err,obj){
 
 ## Test
 
-Tests are executed against a Solr instance in a Docker container. In order to execute them on latest supported Solr version, run:
+Tests are executed against a Solr instance in a Docker container.
```
Comment (kibertoad) [inline README.md:64]: Why remove the clarification?

File: package.json [hunk 2]
```diff
@@ -23,7 +23,6 @@
     "node": ">= 12"
   },
   "dependencies": {
-    "bluebird": "^3.7.2",
```
Comment (kibertoad) [inline package.json:26]: I don't see any changes removing dependency on bluebird, are you sure it is safe to drop it?

File: lib/client.js [hunk 3]
```diff
@@ -52,78 +54,26 @@ function handleJSONResponse(request, bigint, callback) {
...
- * @param {Boolean} params.secure -
- * @param {Boolean} params.bigint -
- * @param {http.Agent} [params.agent] -
- * @param {Function} callback(err,obj) - a function executed when the Solr server responds or an error occurs
- * @param {Error} callback().err
- * @param {Object} callback().obj - JSON response sent by the Solr server deserialized
- *
- * @return {http.ClientRequest}
- * @api private
- */ function postJSON(params, callback) {
```
Comment (kibertoad) [inline lib/client.js:88]: was this not used anywhere?

File: examples/get.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ const solr = require('./../lib/solr');
 const client = solr.createClient();
 
 const query = 'q=id:UTF8TEST&mlt.fl=manu,cat&mlt.mindf=1&mlt.mintf=1';
-client.get('mlt', query, function (err, obj) {
+client.doQuery('mlt', query, function (err, obj) {
```
Comment (kibertoad) [inline examples/get.js:11]: why replace `get` with `doQuery`?

Comment (kibertoad) [reply examples/get.js:11]: I'm not a fan of this change, TBH. In my experience, trying to have one super-generic method for doing calls with any HTTP verb ends up being very messy. It will become even worse after we go to TypeScript, as call signatures for GET, POST/PATCH/PUT and DELETE are very different.

File: lib/solr.js [hunk 5]
```diff
@@ -809,146 +692,218 @@ Client.prototype.createQuery = function () {
...
+ * @api private
+ */
+Client.prototype.doQuery = function(
+  handler,
+  query,
+  callback,
+) {
+  // Construct the string to use as query (GET) or body (POST).
+  let data;
+  if (query instanceof Query || query instanceof Collection) {
```
Comment (kibertoad) [inline lib/solr.js:802]: This kind of highly conditional logic is confusing and very hard to read. I would vote in favour of keeping GET and POST requests separate and invoked explicitly.
Reducing amount of lines of code by itself is not necessarily a win, if that results in a spike of complexity.

**Review by kibertoad — CHANGES_REQUESTED** [review]
Let's talk more about how we want to structure HTTP calls.

**Comment by rudolfbyker** [comment]
Thanks for the detailed review. Continuing tomorrow, hopefully.

File: README.md (hunk 1 above)
Comment (rudolfbyker) [reply README.md:64]: I didn't. I just split the paragraph.

File: examples/get.js (hunk 4 above)
Comment (rudolfbyker) [reply examples/get.js:11]: I did not replace it. I just renamed the function, and then refactored the contents. It's still doing the exact same thing as before.

It's also not "super generic". The original solr-client code has this logic, which I did not want to change, because I'm refactoring the code, not rewriting it:

- Build a get query
- Just before sending it, check if the URL would be too long
- If so, throw it away and build a POST query with the query string in the body (not JSON)
- Send the query

This logic is very specific to solr queries, so the word "query" in "doQuery" refers to a Solr query object, not a generic HTTP query.

File: lib/client.js (hunk 3 above)
Comment (rudolfbyker) [reply lib/client.js:88]: It was, but it was a very large chunk of duplicated code. It is all contained in `doRequest` now. I know you said you do not like the idea of a generic request function, but it's very easy to split again later. The GET and POST requests only differed by one or two lines, which are neatly contained in `if (method === "POST")` now. There are no other methods in use yet in this library. If the requirements diverge later, we can simply split up the `doRequest` method.

**Comment by rudolfbyker** [comment]
The failing tests are baffling me. I think there might be some time-sensitivity in the tests, because consecutive runs of the test suite on my PC will sometimes fail, and sometimes pass.

I will also try to tweak the tests to show the errors instead of just asserting that the error is null and hiding it. At the moment it is clear that some tests are failing with 400 errors, but we cannot see what they are.

**Comment by rudolfbyker** [comment]
Tests are definitely passing on my PC. Do you think we should try to fix the CI in this PR, or disable the checks temporarily and immediately start working on refactoring the tests for the next PR?

**Comment by rudolfbyker** [comment]
> Let's talk more about how we want to structure HTTP calls.

Maybe a separate issue? I know it looks very different now, but this PR does not change any of the HTTP behaviour. It just does the same thing with less code.

File: lib/solr.js (hunk 5 above)
Comment (rudolfbyker) [reply lib/solr.js:802]: This conditional logic is not mine. I just moved it during the refactor. I agree that we should not do this, but I think it's a separate issue.

File: package.json (hunk 2 above)
Comment (rudolfbyker) [reply package.json:26]: There was none, other than a single line: A cal to Promise.promisify, which just adds methods to the prototype. Since we are aiming for a new major version, and I'm going to rewrite all methods as async functions anyway, this does not matter.

File: lib/client.js (hunk 3 above)
Comment (kibertoad) [reply lib/client.js:88]: Sure, let's do it that way then

**Comment by kibertoad** [comment]
@rudolfbyker Considering that tests were consistently not failing before, it really seems that they are doing so now during some of the changes. Breaking CI in master is generally a very bad idea. Maybe we can reduce the scope of this PR and try to identify at which stage they are still passing? (e. g. keep the bluebird or do not yet remove postAsJson, because maybe there is a slight difference in logic)

**Comment by rudolfbyker** [comment]
OK, working on this again today. I will try to split up the PR into smaller parts.
